text/3437-implementable-trait-alias.md

```diff
@@ -0,0 +1,743 @@
+
+use frob_lib::Frob;
+
+struct MyType;
+
+impl Frob for MyType {
+    // The return future of this method is implicitly `Send`, as implied by the alias.
+    // No `#[refine]` is necessary.
+    async fn frob(&self) { /* ... */ }
+}
+```
+
+Trait aliases are `unsafe` to implement iff the underlying trait is marked
+`unsafe`.
+
+## Usage in paths
+
+Implementable trait aliases can also be used with trait-qualified and
+fully-qualified method call syntax, as well as in paths more generally. When
+used this way, they are treated equivalently to the underlying primary trait,
+with the additional restriction that all `where` clauses and type
+parameter/associated type bounds must be satisfied.
+
+```rust
+use std::array;
+
+trait IntIter = Iterator<Item = u32> where Self: Clone;
+
+let iter = [1_u32].into_iter();
+let _: IntIter::Item = IntIter::next(&mut iter); // works
+let _: <array::IntoIter as IntIter>::Item = <array::IntoIter as IntIter>::next(); // works
+//IntIter::clone(&iter); // ERROR: trait `Iterator` has no method named `clone()`
+let dyn_iter: &mut dyn Iterator<Item = u32> = &mut iter;
+//IntIter::next(dyn_iter); // ERROR: `dyn Iterator<Item = u32>` does not implement `Clone`
+let signed_iter = [1_i32].into_iter();
+//IntIter::next(&mut signed_iter); // ERROR: Expected `<Self as Iterator>::Item` to be `u32`, it is `i32`
+```
+
+Implementable trait aliases can also be used with associated type bounds; the
+associated type must belong to the alias's primary trait.
+
+```rust
+trait IteratorAlias = Iterator;
+let _: IteratorAlias<Item = u32> = [1_u32].into_iter();
+
+trait IntIter = Iterator<Item = u32> where Self: Clone;
+let _: IntIter<Item = u32> = [1_u32].into_iter(); // `Item = u32` is redundant, but allowed
+//let _: IntIter<Item = f64> = [1.0_f64].into_iter(); // ERROR: `Item = f64` conflicts with `Item = u32`
+```
+
+# Drawbacks
+
+- The syntactic distance between implementable and non-implementable aliases is
+  short, which might confuse users. In particular, the fact that
+  `trait Foo = Bar + Send;` means something different than
+  `trait Foo = Bar where Self: Send;` will likely be surprising to many.
+- Adds complexity to the language, which might surprise or confuse users.
+- Many of the motivating use-cases involve language features that are not yet
+  stable, or even merely speculative. More experience with those features might
+  unearth better alternatives.
+
+# Rationale and alternatives
+
+- Very lightweight, with no new syntax forms. Compare "trait transformers"
+  proposals, for example—they are generally much heavier.
+  - However, trait transformers would also address more use-cases (for example,
+    sync and async versions of a trait).
+- Better ergonomics compared to purely proc-macro based solutions.
+- One alternative is to allow marker traits or auto traits to appear in `+`
+  bounds of implementable aliases. (For example, `trait Foo = Bar + Send;` could
+  be made implementable).
```

Comment:
I'm in favor of this, because while the symmetry with impl blocks is nice I think the symmetry with bounds is ultimately more important (and more confusing to break). If other lang team members agree I think we should go ahead and support it on day 1.

Otherwise, given that this is something we can allow in the future, can it be listed as a future possibility?